src/autotester/server/utils/tests/file_management_test.py

```diff
@@ -51,167 +64,197 @@ def test_clean_dir_name():
+
+    def test_dir_has_onefile(self, empty_dir):
+        """
+        When the Source Directory has only one file
+        """
+        source_dir = tempfile.mkdtemp()
+        fd, source_file = tempfile.mkstemp(dir=source_dir)
+        dest_dir = empty_dir
+        move_tree(source_dir, dest_dir)
+        moved_file = os.path.join(dest_dir, os.path.basename(source_file))
+        assert os.path.exists(moved_file) and not os.path.exists(source_file)
+
+    def test_dir_has_onedir(self, empty_dir, dir_has_onedir):
+        """
+        When the Dirctory has only one sub directory
+        """
+        source_dir, sub_dir = dir_has_onedir
+        dest_dir = empty_dir
+        move_tree(source_dir, dest_dir)
+        moved_dir = os.path.join(dest_dir, os.path.basename(sub_dir))
+        assert os.path.exists(moved_dir) and not os.path.exists(sub_dir)
+
+    def test_dir_has_nestedfd(self, empty_dir, nested_fd):
+        """
+        When the files are nested in subdirectories more than 2 directories deep
+        """
+        dest_dir = empty_dir
+        source_dir = tempfile.mkdtemp()
+        sub_dir1 = tempfile.mkdtemp(dir=source_dir)
+        sub_dir2 = tempfile.mkdtemp(dir=sub_dir1)
+        fd, source_file = tempfile.mkstemp(dir=sub_dir2)
+        move_tree(source_dir, dest_dir)
+        moved_dir1 = os.path.join(dest_dir, os.path.basename(sub_dir1))
+        moved_dir2 = os.path.join(moved_dir1, os.path.basename(sub_dir2))
+        moved_file = os.path.join(moved_dir2, os.path.basename(source_file))
+        assert os.path.exists(moved_dir1) and not os.path.exists(sub_dir1)
+        assert os.path.exists(moved_dir2)
+        assert os.path.exists(moved_file)
+
+
+class TestFdOpen:
+    def test_open_dir(self):
+        with tempfile.TemporaryDirectory() as dir:
+            dir_fd = os.open(dir, os.O_RDONLY)
+            with fd_open(dir) as fdd:
+                same_dir = os.path.sameopenfile(fdd, dir_fd)
+            assert same_dir
+
+    def test_open_file(self):
+        with tempfile.NamedTemporaryFile() as file:
+            file_fd = os.open(file.name, os.O_RDONLY)
+            with fd_open(file.name) as fdf:
+                same_file = os.path.sameopenfile(fdf, file_fd)
+            assert same_file
+
+    def test_close(self):
+        with tempfile.TemporaryDirectory() as dir:
+            with fd_open(dir) as fdd:
+                dir_fd = fdd
+            with pytest.raises(Exception):
+                assert os.close(dir_fd)
+
+
+def test_fd_lock(dir_has_onefile, empty_dir):
+    temp_dir, temp_file = dir_has_onefile
+    dest_dir = empty_dir
     access = True
     with fd_open(temp_dir) as fd:
         with fd_lock(fd):
-            try:
-                subprocess.run(["rm", {temp_file}])
-            except Exception:
-                access = False
-    assert not access
+            with pytest.raises(Exception) as excinfo:
+                assert subprocess.run(["cp", temp_file, dest_dir])
```

Comment:
You don't want to assert anything here. You're just running the subprocess command and seeing if it raises the error you expect.

Also, before you were doing `["rm", {temp_file}]` and now you're doing `["cp", temp_file, dest_dir]` is that intended?

Reply:
No, Just tried different commands.

Reply:
Maybe I should have brought this up earlier... fcntl only enforces advisory locks